collect_app/src/test/java/org/odk/collect/android/formentry/QuitFormDialogFragmentTest.java

```diff
@@ -0,0 +1,54 @@
+package org.odk.collect.android.formentry;
+
+import android.app.Dialog;
+
+import androidx.fragment.app.FragmentActivity;
+import androidx.fragment.app.FragmentManager;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.Robolectric;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.LooperMode;
+import org.robolectric.shadows.ShadowDialog;
+
+import static junit.framework.TestCase.assertTrue;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.Assert.assertFalse;
+import static org.robolectric.Shadows.shadowOf;
+
+@RunWith(RobolectricTestRunner.class)
+@LooperMode(LooperMode.Mode.PAUSED)
+public class QuitFormDialogFragmentTest {
+
+    private FragmentManager fragmentManager;
+    private QuitFormDialogFragment dialogFragment;
+
+    @Before
+    public void setup() {
+        FragmentActivity activity = Robolectric.setupActivity(FragmentActivity.class);
+        fragmentManager = activity.getSupportFragmentManager();
+        dialogFragment = new QuitFormDialogFragment();
+    }
+
+    @Test
+    public void dialogIsCancellable() {
+        dialogFragment.show(fragmentManager, "tag");
+
+        fragmentManager.executePendingTransactions();
+        assertThat(shadowOf(dialogFragment.getDialog()).isCancelable(), equalTo(true));
+    }
+
+    @Test
+    public void dismiss_shouldDismissTheDialog() {
```

Comment:
Do we need this test? I think there are a few things that we could add tests for here but `dismiss` working feels like it's the framework's responsibility. Maybe I'm missing something here?

Reply:
Dismissing the dialog on dismissing the fragment is the framework responsibility. I can think of one case if I override the `onDismiss` method and do not call super.onDestroy, it retains the dialog even when I call `dismiss`. 
It would be more interesting to add tests to check whether the dialog is retained during orientation change, when the fragment is recreated.

Reply:
That is good thinking! I think it's potentially overkill though. I think it's best to think about tests protecting from "lazy" programmers (as we're all pretty lazy really) - protect from things being deleted or changed but not added. The universe of things that someone could add are pretty big and so it's hard to test protectively against it. Does that make sense?

Reply:
Yes @seadowg, I think it's better if I add the test to check whether the dialog is dismissed when the user clicks the positive button. And to check whether the view is populated correctly and all the options are working or not. I think that will verify the basic functionality of the dialog.